allennlp/layers/embeddings.py

```diff
@@ -0,0 +1,173 @@
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            self.reset_parameters()
+
+        else:
+            if weight.size() != (num_embeddings, embedding_dim):
+                raise ConfigurationError("A weight matrix was passed with contradictory embedding shapes.")
+            self.weight = torch.nn.Parameter(weight, requires_grad=trainable)
+            if self.padding_index is not None:
+                self.weight.data[self.padding_index].fill_(0)
+
+    def reset_parameters(self):
+        self.weight.data.normal_(0, 1)
+        if self.padding_index is not None:
+            self.weight.data[self.padding_index].fill_(0)
+
+    def forward(self, input):
+        padding_idx = self.padding_idx
+        if padding_idx is None:
+            padding_idx = -1
+        return self._backend.Embedding(padding_idx,
+                                       self.max_norm,
+                                       self.norm_type,
+                                       self.scale_grad_by_freq,
+                                       self.sparse)(input, self.weight)
+
+
+def get_pretrained_embedding_layer(embeddings_filename: str,
+                                   vocab: Vocabulary,
+                                   namespace: str="tokens",
+                                   trainable: bool=True,
+                                   log_misses: bool=False):
+    """
+    Reads a pre-trained embedding file and generates a Keras Embedding layer that has weights
+    initialized to the pre-trained embeddings.  The Embedding layer can either be trainable or
+    not.
+
+    We use the DataIndexer to map from the word strings in the embeddings file to the indices
+    that we need, and to know which words from the embeddings file we can safely ignore.  If we
+    come across a word in DataIndexer that does not show up with the embeddings file, we give
+    it a zero vector.
+
+    The embeddings file is assumed to be gzipped, formatted as [word] [dim 1] [dim 2] ...
+    """
+    words_to_keep = set(vocab.get_index_to_token_vocabulary(namespace).values())
+    vocab_size = vocab.get_vocab_size(namespace)
+    embeddings = {}
+    embedding_dim = None
+
+    # TODO(matt): make this a parameter
+    embedding_misses_filename = 'embedding_misses.txt'
+
+    # First we read the embeddings from the file, only keeping vectors for the words we need.
+    logger.info("Reading embeddings from file")
+    with gzip.open(embeddings_filename, 'rb') as embeddings_file:
+        for line in embeddings_file:
+            fields = line.decode('utf-8').strip().split(' ')
+            if embedding_dim is None:
+                embedding_dim = len(fields) - 1
+                assert embedding_dim > 1, "Found embedding size of 1; do you have a header?"
+            else:
+                if len(fields) - 1 != embedding_dim:
+                    # Sometimes there are funny unicode parsing problems that lead to different
```

Comment:
Can we do better here?  Maybe we can just import a library that reads glove vectors already, so we don't have to duplicate that functionality?